### PR TITLE
Update source-generators.cookbook.md for recommended testing approach

### LINKED ISSUE #50902: Clarify testing recommendations for source generators
As a developer trying to write unit tests for source generators, the existing docs lead me astray.

- The Source Generators Cookbook describes an outdated approached for testing source generators: https://github.com/dotnet/roslyn/blob/master/docs/features/source-generators.cookbook.md#unit-testing-of-generators
- The testing package (https://github.com/dotnet/roslyn-sdk/tree/master/src/Microsoft.CodeAnalysis.Testing) doesn't reference the new `Microsoft.CodeAnalysis.CSharp.SourceGenerators.Testing` which is the recommended approach.

Some things that would help:

- Updating the sample in the cookboook.
- Making the `Microsoft.CodeAnalysis.CSharp.SourceGenerators.Testing` more discoverable.

### PR DESCRIPTION
**Pull request by Youssef1313**
Haven't tried these instructions myself, but taken from @sharwell's PR in https://github.com/dotnet/roslyn-analyzers/pull/4897/

Fixes #50902

### CODE REVIEW

File: docs/features/source-generators.cookbook.md [hunk 1]
```diff
@@ -624,105 +624,77 @@ public class MyGenerator : ISourceGenerator
...
+And use the following in your test method:
+
+```csharp
+var code = "initial code"
+var generated = "expected generated code";
+await new VerifyCS.Test
+{
+    TestCode = code,
+    FixedCode = generated,
+}.RunAsync();
```
[the remark this answers is not in the document]
Comment (Youssef1313) [reply docs/features/source-generators.cookbook.md]: Good catch. Thanks @naminodarie. I updated the code, does it look okay now?

**Comment by chsienki** [comment]
@Youssef1313 apologies, this dropped off of my radar.

The new samples are great, but I think I'd also like to keep the existing unit test advice, as not everyone is going to want to use the full fledged verifier framework. Could you split the change out into adding the new bits, and giving the two recipes a sensible name (like Unit Test directly, and Unit test via Verification, or something?)

Thanks!

**Comment by sharwell** [comment]
@chsienki I would accept a compromise approach of listing both testing strategies on two conditions:

1. The Microsoft.CodeAnalysis.Testing approach is listed first
2. The manual approach starts with a ⚠ clear call-out saying it is not the recommended testing strategy for source generator authors, but remains listed to serve as documentation for how the backing APIs work

If this doesn't work for you we should set up a meeting to discuss options.

File: docs/features/source-generators.cookbook.md [hunk 2]
```diff
@@ -624,105 +624,83 @@ public class MyGenerator : ISourceGenerator
...
+{
+    TestState = 
+    {
+        Sources = { code },
+    },
+    FixedState = 
+    {
+        Sources = { code, generated },
+    }
+}.RunAsync();
```
Comment (sharwell) [inline docs/features/source-generators.cookbook.md:702]: This was rewritten in https://github.com/dotnet/roslyn-analyzers/commit/8b0215e8b67e4d10695ad7d2a0cfd7e210cad869

**Comment by jaredpar** [comment]
> The manual approach starts with a ⚠ clear call-out saying it is not the recommended testing strategy for source generator authors, but remains listed to serve as documentation for how the backing APIs work

Fundamentally do not agree with this. There is nothing wrong with writing tests the old way (customers and ourselves have been doing it this way for years). There are plenty of established patterns for approaching this that are in high use. The existence of the test framework does not render these invalid nor does it warrant a caution here.

File: docs/features/source-generators.cookbook.md [hunk 3]
```diff
@@ -624,7 +624,86 @@ public class MyGenerator : ISourceGenerator
 
 **User scenario**: As a generator author, I want to be able to unit test my generators to make development easier and ensure correctness.
 
-**Solution**: A user can host the `GeneratorDriver` directly within a unit test, making the generator portion of the code relatively simple to unit test. A user will need to provide a compilation for the generator to operate on, and can then probe either the resulting compilation, or the `GeneratorDriverRunResult` of the driver to see the individual items added by the generator.
+**Solution A**:
+
+The recommended approach is to use [Microsoft.CodeAnalaysis.Testing](https://github.com/dotnet/roslyn-sdk/tree/main/src/Microsoft.CodeAnalysis.Testing#microsoftcodeanalysistesting) packages:
```
Comment (jcouv) [inline docs/features/source-generators.cookbook.md:629]:
```suggestion
The recommended approach is to use [Microsoft.CodeAnalysis.Testing](https://github.com/dotnet/roslyn-sdk/tree/main/src/Microsoft.CodeAnalysis.Testing#microsoftcodeanalysistesting) packages:
```

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 4) modulo typo

**Comment by jmarolf** [comment]
> There is nothing wrong with writing tests the old way (customers and ourselves have been doing it this way for years)

This is most certainly true, but doing this requires a lot of knowledge of how roslyn works to set up correctly. The hope is that for common scenarios this is abstracted away.

**Review by jmarolf — APPROVED** [review]

**Comment by davidwengier** [comment]
> Fundamentally do not agree with this.

I fundamentally agree with the fundamental not agreeing. I've pointed a few people on the C# Discord at the testing framework when they come across my "old way" tests, and the feedback is usually "interesting, but overkill for my needs."

If the documentation in the Roslyn repo can't use the Roslyn API then the API probably needs to change. The testing framework is a layer on top, and absolutely should be included (our documentation in general for source generators could use all of the help it can get) but it's most certainly not the only way. I'm not sure the document should even say "recommended" without being clear who its recommended by and in what scenarios.

**Comment by sharwell** [comment]
> the feedback is usually "interesting, but overkill for my needs."

The test library hides a tremendous amount of complexity. It is unlikely that a person would meet both of the following conditions:

1. They understand the testing requirements for solid regression prevention in an analyzer or source generator
2. They want to implement the test infrastructure themselves and/or do not want to use the test library

**Comment by davidwengier** [comment]
Broadly speaking, I feel like I fit into those criteria. To rule me (and probably others) out, I'd have to add a third: "They don't take a pragmatic approach to software development."

I'll happily put my hand up and say I'm not using the testing framework, however that is not because I think what I have now is at the same level, rather it's because I understand the appropriate level I need for my projects. Hell I even know some bugs in my source generators that I'm not testing, and not fixing, because I understand the likelihood of them being hit.

Not everyone needs or wants the same thing and I don't think we should be warning people off something that might be entirely appropriate for them.

> It took a couple years of fighting real regressions in analyzers to develop the library in a manner that's capable of preventing them in the first place.

If it seems like I'm speaking ill of the testing framework I apologize profusely, that is not my intention. I just think we need to be honest and recognize that it is not the only option, and that the need for it in some projects does not rule out or invalidate other options.

**Comment by sharwell** [comment]
I did not interpret your comments as speaking ill of the test framework, and even if that was your intent I would be OK with that. 😄 My advocacy of a specific approach is less about pushing something I happened to create and more about pointing someone towards a set of tools which, in my experience, happens to do the best job of ensuring and maintaining software quality with the least amount of total invested time.

**Comment by jcouv** [comment]
@Youssef1313 As far as I can tell, this is ready to merge. Anything left from your side?

**Comment by Youssef1313** [comment]
> Anything left from your side

Nothing left if the team agrees on documenting both approaches. Thanks!

**Comment by jcouv** [comment]
The PR is a clear improvement over previous doc. I'll go ahead with merge and we can always revise further. Thanks

**Pull request merged**